vizro-core/src/vizro/managers/_data_manager.py

```diff
@@ -197,18 +162,35 @@ def __init__(self):
-                    "code to use a full function, e.g. using `lambda`."
-                )
-            self.__data[name] = _DynamicData(name, data)
+            # __qualname__ is required by flask-caching (even if we specify our own make_name) but
+            # not defined for partial functions and just '<lambda>' for lambda functions. Defining __qualname__
+            # means it's possible to have non-interfering caches for lambda functions (similarly if we
+            # end up using CapturedCallable, or that could instead set its own __qualname__).
+            # We handle __name__ the same way even though it's not currently essential to functioning of flask-caching
+            # in case they change the underlying implementation to use it.
+            # We use partial to effectively make an independent copy of the underlying data function. This means that
+            # it's possible to set __qualname__ independently for each data source. This is not essential for
+            # functions other than lambda, but it is essential for bound methods, as flask-caching cannot easily
+            # independently timeout different instances with different bound methods but the same underlying function
+            # data.__func__. If we don't do this then the bound method case needs some uglier hacking to make work
+            # correctly - see https://github.com/mckinsey/vizro/blob/abb7eebb230ba7e6cfdf6150dc56b211a78b1cd5/
+            # vizro-core/src/vizro/managers/_data_manager.py.
+            # Once partial has been used, all dynamic data sources are on equal footing since they're all treated as
+            # functions rather than bound methods, e.g. by flask_caching.utils.function_namespace. This makes it much
+            # simpler to use flask-caching reliably.
+            # It's important the __qualname__ is the same across all workers, so use the data source name rather than
+            # e.g. the repr method that includes the id of the instance so would only work in the case that gunicorn is
+            # running with --preload.
+            # __module__ is also required in flask_caching.utils.function_namespace and not defined for partial
+            # functions in some versions of Python.
+            # update_wrapper ensures that __module__, __name__, __qualname__, __annotations__ and __doc__ are
+            # assigned to the new partial(data) the same as they were in data. This isn't strictly necessary but makes
+            # inspecting these functions easier.
+            data = functools.update_wrapper(partial(data), data)
+            data.__module__ = getattr(data, "__module__", "<nomodule>")
+            data.__name__ = ".".join([getattr(data, "__name__", "<unnamed>"), name])
+            data.__qualname__ = ".".join([getattr(data, "__qualname__", "<unnamed>"), name])
+            self.__data[name] = _DynamicData(data)
```

Comment:
I tried this out in https://github.com/mckinsey/vizro/pull/479/commits/191868fcda33a36a388dd2bee2e1ea75bef3afa6 and it almost works but not quite.

Rewriting the function signature is not too bad since it's very similar to [this wrapt example](https://wrapt.readthedocs.io/en/latest/decorators.html#signature-changing-decorators). The biggest catch is actually associated with a note I added to one of the tests:
> When the cache for one set of memoized arguments expires, the cache for the whole data source expires, even for other values of memoized arguments.

This means that `test_independent_dynamic_data_callable_with_timeout` starts failing, because if you have the following then the `timeout` setting applies to both `data_x` and `data_y`.
```
data_manager["data_x"] = RandomData().load
data_manager["data_y"] = RandomData().load
data_manager["data_y"].timeout = 1
```

This is a big problem when it comes to kedro datasets which all work according to this pattern.

This means that for the case of bound methods we still need to pass in functions rather than bound methods, and those functions still need to be labelled according to the underlying data source name. It's also a problem for `lambda` and `partial` functions. We also still need to assign a `__qualname__` anyway in order to be able to handle `partial` function at alls, which is [something we probably do want to be able to do](https://github.com/mckinsey/vizro/pull/451#issuecomment-2111138606). So most of the ugly `__qualname__` stuff is still needed anyway.

Hence I've reverted trying this out. But thank you for the excellent comment, it was very useful to investigate!